modules/shortcodes/gist.php

```diff
@@ -22,11 +34,73 @@
  * @param array  $rawattr The original unmodified attributes.
  * @return string The embed HTML.
  */
-function github_gist_embed_handler( $matches, $attr, $url, $rawattr ) {
-	// Let the shortcode callback do all the work
+function github_gist_embed_handler( $matches, $attr, $url, $rawattr ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+	// Let the shortcode callback do all the work.
 	return github_gist_shortcode( $matches, $url );
 }
 
+/**
+ * Extract an ID from a Gist shortcode or a full Gist URL.
+ *
+ * @since 7.3.0
+ *
+ * @param string $gist Gist shortcode or full Gist URL.
+ *
+ * @return array $gist_info {
+ * Array of information about our gist.
+ *     @type string $id   Unique identifier for the gist.
+ *     @type string $file File name if the gist links to a specific file.
+ * }
+ */
+function jetpack_gist_get_shortcode_id( $gist = '' ) {
+	$gist_info = array(
+		'id'   => '',
+		'file' => '',
+	);
+	// Simple shortcode, with just an ID.
+	if ( ctype_alnum( $gist ) ) {
+		$gist_info['id'] = $gist;
+	}
+
+	// Full URL? Only keep the relevant parts.
+	$parsed_url = wp_parse_url( $gist );
+	if (
+		! empty( $parsed_url )
+		&& is_array( $parsed_url )
+		&& isset( $parsed_url['scheme'], $parsed_url['host'], $parsed_url['path'] )
+	) {
+		// Not a Gist URL? Bail.
+		if ( 'gist.github.com' !== $parsed_url['host'] ) {
+			return array(
+				'id'   => '',
+				'file' => '',
+			);
+		}
+
+		// Keep the file name if there was one.
+		if ( ! empty( $parsed_url['fragment'] ) ) {
+			$gist_info['file'] = preg_replace( '/(?:file-)(.+)/', '$1', $parsed_url['fragment'] );
+		}
+
+		// Keep the unique identifier without any leading or trailing slashes.
+		if ( ! empty( $parsed_url['path'] ) ) {
+			$gist_info['id'] = preg_replace( '/(?:\/)([^\.]+)\./', '$1', $parsed_url['path'] );
+			// Overwrite $gist with our identifier to clean it up below.
+			$gist = $gist_info['id'];
+		}
+	}
+
+	// Not a URL nor an ID? Only keep one of "username/id", "/username/id", "id".
+	if ( preg_match( '#^/?(([a-z0-9-_]+/)?([a-z0-9]+))$#i', $gist, $matches ) ) {
```

Comment:
It doesn't matter in this regex, but it's usually best to put any `-` that is not part of a range at the end of the character class so that it cannot be interpreted as a range character.

```suggestion
	if ( preg_match( '#^/?(([a-z0-9_-]+/)?([a-z0-9]+))$#i', $gist, $matches ) ) {
```